math_utils.py

```diff
@@ -74,12 +78,25 @@ def PLL(Onsets,fs): ## TODO: Implement
     for n in range(len(Onsets)-1):
         theta[n] = phase_comp(Onsets[n],x[n])
         e[n] = loop_filter(theta,n,loop_filt_type,T=loop_filter_mem)     
-        x_phase[n+1] = VCO(e[n],x_phase[n], VCO_gain)       
-        x[n+1] = np.sin(np.unwrap([2*np.pi*f0*n/fs+x_phase[n+1]]))
-        # x[t+1] = np.sin(np.unwrap([x_phase[t+1]]))
-        
+        x_phase[n+1] = VCO(e[n],x_phase[n], VCO_gain) 
+        inst_f[n+1]  = f0 + + (x_phase[n+1] - x_phase[n])*fs/(2*np.pi)    
+        x[n+1] = np.sin(np.unwrap([2*np.pi*f0*n/fs+x_phase[n+1]]))       
         if(n%10000==0): ## For debug
             print(n/10000)
-    return x
+    return e, x
+
+def zero_cross_detect(x):
+    zc_vec = np.zeros(x.shape)
+    zc_vec[x>0] = 1
+    zc_vec=np.diff(zc_vec)
+    zc_vec[zc_vec<0]=0
+    return zc_vec
+
+def metronome_thresholding(x,th=0.999):
+    metronome = np.zeros(x.shape)
+    metronome[x>th]=1
+    return metronome
+    
+    
```

Comment:
Do we need both?

Reply:
No, I think the second is simple and work well. 
We only might need to synchronize the metronome and s_t. I have an idea how to do it